File: backend/utils/file_router/model_matcher.py
"""
Model Matcher - Intelligent model-file compatibility checker
Analyzes which models can handle specific file types
"""

from typing import Dict, List, Any, Optional


class ModelMatcher:
    """
    Matches files with compatible AI models based on capabilities
    """
    
    # Import model categories from existing model_router
    # This will be integrated with the actual MODEL_CATEGORIES
    
    @staticmethod
    def get_compatible_models(file_type: str, file_category: str) -> Dict[str, Any]:
        """
        Get list of models compatible with given file type
        
        Args:
            file_type: Type of file (pdf, image, docx, pptx)
            file_category: Category (document, image)
            
        Returns:
            Dict with compatible models and recommendations
        """
        # For PDFs - All models support (DeepSeek native, others with extraction)
        if file_type == 'pdf':
            return {
                'compatible_models': 'all',  # All models can handle PDFs
                'recommended_model': 'deepseek-chat-v3.2-exp',  # Best for PDFs
                'reason': 'DeepSeek models support native PDF reading with 128K context. Other models will extract text.',
                'max_files': 3,  # DeepSeek limit; general models limited to 1 by validation
                'supports_native': False,  # Only DeepSeek has native support
                'category': 'pdf_analysis',
            }
        
        # For images - NOT SUPPORTED (vision models removed)
        elif file_type == 'image':
            return {
                'compatible_models': [],
                'recommended_model': None,
                'reason': 'Image analysis not supported. Vision models have been disabled.',
                'max_files': 0,
                'supports_native': False,
                'category': 'unsupported',
            }
        
        # For DOCX/PPTX - all models support after text extraction
        # NOTE: This returns DEFAULT values. Actual limits should be checked
        # via ValidationRules which is model-aware (DeepSeek=3, others=1)
        elif file_type in ['docx', 'pptx']:
            return {
                'compatible_models': 'all',  # All models support text
                'recommended_model': 'deepseek-chat-v3.2-exp',  # Best for multi-docs
                'reason': 'Text will be extracted and sent to any model. DeepSeek supports up to 3 files.',
                'max_files': 3,  # Default to DeepSeek's limit; will be validated per model
                'supports_native': False,
                'category': 'general',  # Backend will map to pdf_analysis for DeepSeek
            }
        
        return {
            'compatible_models': [],
            'recommended_model': None,
            'reason': 'Unsupported file type',
            'max_files': 0,
            'supports_native': False,
            'category': 'unknown',
        }
    
    @staticmethod
    def check_model_compatibility(
        current_model: str,
        file_type: str,
        file_category: str,
        file_count: int = 1
    ) -> Dict[str, Any]:
        """
        Check if current model is compatible with the file(s)
        
        Args:
            current_model: Currently selected model ID
            file_type: Type of file being uploaded
            file_category: Category of file
            file_count: Number of files being uploaded
            
        Returns:
            Dict with compatibility status and suggestions
        """
        compatibility_info = ModelMatcher.get_compatible_models(file_type, file_category)
        
        # Extract model family from full model ID
        model_family = ModelMatcher._get_model_family(current_model)
        
        # Check if current model is compatible
        compatible_models = compatibility_info['compatible_models']
        is_compatible = (
            compatible_models == 'all' or
            current_model in compatible_models or
            model_family in ['deepseek']  # Removed vision models
        )
        
        # Get model-specific file count limits
        max_files = ModelMatcher._get_max_files_for_model(current_model, file_type)
        exceeds_limit = file_count > max_files
        
        return {
            'is_compatible': is_compatible and not exceeds_limit,
            'current_model': current_model,
            'file_type': file_type,
            'file_count': file_count,
            'max_files': max_files,
            'exceeds_limit': exceeds_limit,
            'recommended_model': compatibility_info['recommended_model'],
            'compatible_models': compatible_models,
            'reason': compatibility_info['reason'],
            'action_required': not is_compatible or exceeds_limit,
        }
    
    @staticmethod
    def _get_max_files_for_model(model_id: str, file_type: str) -> int:
        """
        Get maximum file count for specific model and file type
        Model-aware limits: DeepSeek supports 3 PDFs/DOCX/PPTX, Vision supports 10 images
        
        Args:
            model_id: Model identifier
            file_type: Type of file
            
        Returns:
            Maximum file count for this model-filetype combination
        """
        model_family = ModelMatcher._get_model_family(model_id)
        
        # DeepSeek models support 3 PDFs, 3 DOCX, 3 PPTX
        if model_family == 'deepseek':
            if file_type in ['pdf', 'docx', 'pptx']:
                return 3
        
        # Vision models support 10 images - REMOVED (vision models disabled)
        # Default: 1 file for all other combinations
        return 1
    
    @staticmethod
    def _get_model_family(model_id: str) -> str:
        """
        Extract model family from full model ID
        
        Examples:
            'deepseek-chat-v3.2-exp' -> 'deepseek'
            'provider-6/gemma-3-27b-instruct' -> 'gemma'
            'provider-6/llama-3.2-11b-instruct' -> 'llama'
        """
        model_lower = model_id.lower()
        
        if 'deepseek' in model_lower:
            return 'deepseek'
        elif 'qwen' in model_lower:
            return 'qwen'
        elif 'llama' in model_lower:
            return 'llama'
        elif 'gemma' in model_lower:
            return 'gemma'
        elif 'gpt' in model_lower:
            return 'gpt'
        elif 'mistral' in model_lower:
            return 'mistral'
        
        return 'unknown'
    
    @staticmethod
    def get_model_capabilities(model_id: str) -> Dict[str, Any]:
        """
        Get capabilities of a specific model
        
        Args:
            model_id: Model identifier
            
        Returns:
            Dict with model capabilities
        """
        model_family = ModelMatcher._get_model_family(model_id)
        
        if model_family == 'deepseek':
            return {
                'supports_pdf': True,
                'supports_images': False,
                'max_pdf_count': 3,
                'context_window': 128000,
                'native_pdf_support': True,
            }
        elif model_family in ['gemma', 'llama']:  # Removed vision-specific handling
            return {
                'supports_pdf': False,
                'supports_images': False,
                'max_pdf_count': 1,
                'context_window': 32000,
                'native_pdf_support': False,
            }
        else:
            return {
                'supports_pdf': False,
                'supports_images': False,
                'max_pdf_count': 1,
                'context_window': 8000,
                'native_pdf_support': False,
            }

File: frontend/utils/fileRouter/modelMatcher.ts
/**
 * Model Matcher - Frontend utility for model-file compatibility
 * Mirrors backend ModelMatcher logic for client-side checks
 */

export interface ModelCompatibility {
  compatibleModels: string[] | 'all';
  recommendedModel: string | null;
  reason: string;
  maxFiles: number;
  supportsNative: boolean;
  category: string;
}

export interface CompatibilityCheck {
  isCompatible: boolean;
  currentModel: string;
  fileType: string;
  fileCount: number;
  maxFiles: number;
  exceedsLimit: boolean;
  recommendedModel: string | null;
  compatibleModels: string[] | 'all';
  reason: string;
  actionRequired: boolean;
}

/**
 * Get compatible models for a file type
 * NOTE: This returns DEFAULT compatibility. For model-specific limits, 
 * use getMaxFilesForModelAndType() or call backend API
 */
export function getCompatibleModels(
  fileType: string,
  fileCategory: string
): ModelCompatibility {
  // For PDFs
  if (fileType === 'pdf') {
    return {
      compatibleModels: [
        'deepseek-chat-v3.2-exp',
        'deepseek-chat-v3.2-coder',
        'deepseek-reasoner-r1',
      ],
      recommendedModel: 'deepseek-chat-v3.2-exp',
      reason: 'DeepSeek models support native PDF reading with 128K context',
      maxFiles: 3,
      supportsNative: true,
      category: 'pdf_analysis',
    };
  }

  // For images
  if (fileType === 'image') {
    return {
      compatibleModels: ['pixtral-12b', 'llama-3.2-vision-90b'],
      recommendedModel: 'pixtral-12b',
      reason: 'Vision models can analyze images directly',
      maxFiles: 10,
      supportsNative: true,
      category: 'vision',
    };
  }

  // For DOCX/PPTX - DeepSeek supports 3, others support 1
  if (fileType === 'docx' || fileType === 'pptx') {
    return {
      compatibleModels: 'all',
      recommendedModel: 'deepseek-chat-v3.2-exp',
      reason: 'Text will be extracted and sent to any model. DeepSeek supports up to 3 files.',
      maxFiles: 3, // Default to DeepSeek's limit, will be adjusted per model
      supportsNative: false,
      category: 'general', // Note: Backend will map to pdf_analysis for DeepSeek
    };
  }

  return {
    compatibleModels: [],
    recommendedModel: null,
    reason: 'Unsupported file type',
    maxFiles: 0,
    supportsNative: false,
    category: 'unknown',
  };
}

/**
 * Check if current model is compatible with file
 */
export function checkModelCompatibility(
  currentModel: string,
  fileType: string,
  fileCategory: string,
  fileCount: number = 1
): CompatibilityCheck {
  const compatibilityInfo = getCompatibleModels(fileType, fileCategory);
  const modelFamily = getModelFamily(currentModel);

  // Check if current model is compatible
  const compatibleModels = compatibilityInfo.compatibleModels;
  const isCompatible =
    compatibleModels === 'all' ||
    compatibleModels.includes(currentModel) ||
    ['deepseek', 'pixtral', 'llama-vision'].includes(modelFamily);

  // Check file count limits - use model-specific limits!
  const maxFiles = getMaxFilesForModelAndType(currentModel, fileType);
  const exceedsLimit = fileCount > maxFiles;

  return {
    isCompatible: isCompatible && !exceedsLimit,
    currentModel,
    fileType,
    fileCount,
    maxFiles,
    exceedsLimit,
    recommendedModel: compatibilityInfo.recommendedModel,
    compatibleModels,
    reason: compatibilityInfo.reason,
    actionRequired: !isCompatible || exceedsLimit,
  };
}

/**
 * Extract model family from model ID
 */
export function getModelFamily(modelId: string): string {
  const modelLower = modelId.toLowerCase();

  if (modelLower.includes('deepseek')) return 'deepseek';
  if (modelLower.includes('pixtral')) return 'pixtral';
  if (modelLower.includes('llama') && modelLower.includes('vision'))
    return 'llama-vision';
  if (modelLower.includes('qwen')) return 'qwen';
  if (modelLower.includes('llama')) return 'llama';
  if (modelLower.includes('gemma')) return 'gemma';

  return 'unknown';
}

/**
 * Get model capabilities
 */
export function getModelCapabilities(modelId: string) {
  const modelFamily = getModelFamily(modelId);

  if (modelFamily === 'deepseek') {
    return {
      supportsPdf: true,
      supportsImages: false,
      maxPdfCount: 3,
      contextWindow: 128000,
      nativePdfSupport: true,
    };
  }

  if (modelFamily === 'pixtral' || modelFamily === 'llama-vision') {
    return {
      supportsPdf: false,
      supportsImages: true,
      maxImageCount: 10,
      contextWindow: 32000,
      nativeImageSupport: true,
    };
  }

  return {
    supportsPdf: false,
    supportsImages: false,
    maxPdfCount: 1,
    contextWindow: 8000,
    nativePdfSupport: false,
  };
}

/**
 * Check if model supports PDF
 */
export function modelSupportsPdf(modelId: string): boolean {
  return getModelFamily(modelId) === 'deepseek';
}

/**
 * Check if model supports images
 */
export function modelSupportsImages(modelId: string): boolean {
  const family = getModelFamily(modelId);
  return family === 'pixtral' || family === 'llama-vision';
}

/**
 * Get max files for model and file type
 */
export function getMaxFilesForModel(
  modelId: string,
  fileType: string
): number {
  const capabilities = getModelCapabilities(modelId);
  const modelFamily = getModelFamily(modelId);

  if (fileType === 'pdf' && capabilities.supportsPdf) {
    return capabilities.maxPdfCount || 1;
  }

  if (fileType === 'image' && capabilities.supportsImages) {
    return capabilities.maxImageCount || 1;
  }

  // For DOCX/PPTX - DeepSeek supports 3, others support 1
  if (fileType === 'docx' || fileType === 'pptx') {
    return modelFamily === 'deepseek' ? 3 : 1;
  }

  return 1; // Default for other file types
}

/**
 * Get max files for model and file type with model-aware logic
 * This version properly handles DeepSeek's multi-file support for DOCX/PPTX
 */
export function getMaxFilesForModelAndType(
  modelId: string,
  fileType: string
): number {
  const modelFamily = getModelFamily(modelId);

  // DeepSeek supports 3 PDFs, 3 DOCX, 3 PPTX
  if (modelFamily === 'deepseek') {
    if (fileType === 'pdf' || fileType === 'docx' || fileType === 'pptx') {
      return 3;
    }
  }

  // Vision models support 10 images
  if (
    (modelFamily === 'pixtral' || modelFamily === 'llama-vision') &&
    fileType === 'image'
  ) {
    return 10;
  }

  // Default: 1 file for all other combinations
  return 1;
}
